### PR TITLE
Add 'deployment.keys.<key>.keyFile' option

### LINKED ISSUE #4: Add support for the 'keyFile' parameter for deployment keys
In the spirit of #3. I am very interested in uploading the key from a local path specified by `keyFile` option ([reference code in NixOps](https://github.com/NixOS/nixops/blob/1ed5a091bc52de6c91319f446f833018a1cb326e/nix/keys.nix#L54-L73)).

@zhaofengli  I am very willing to implement this myself in the coming days, so just creating this issue to inquire whether you might already be on this issue - don't want to duplicate effort.

### PR DESCRIPTION
**Pull request by justinas**
Allows uploading the contents of a local file as a key. Fixes #4.

### CODE REVIEW

File: src/nix/key.rs [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
 use regex::Regex;
-use serde::{Serialize, Deserialize};
+use serde::{Deserialize, Serialize};
+use tokio::{fs::File, io::AsyncRead};
 use validator::{Validate, ValidationError};
 
 #[derive(Debug, Clone, Validate, Serialize, Deserialize)]
 pub struct Key {
-    pub(crate) text: String,
+    pub(crate) text: Option<String>,
```
Comment (justinas) [inline src/nix/key.rs:13]: Could be a proper enum like:
```
Text(String),
KeyFile(String),
```
 but we would have to do more magic with serde

File: src/nix/host/local.rs [hunk 2]
```diff
@@ -149,7 +150,13 @@ impl Local {
 
         let parent_dir = dest_path.parent().unwrap();
         fs::create_dir_all(parent_dir)?;
-        fs::rename(temp_path, dest_path)?;
+
+        if fs::rename(&temp_path, &dest_path).is_err() {
+            // Linux can not rename across different filesystems, try copy-then-remove
```
Comment (justinas) [inline src/nix/host/local.rs:155]: Noticed this when I tried to deploy secrets locally to `/run/keys`, which is a separate tmpfs mount.

File: src/nix/host/local.rs [hunk 3]
```diff
@@ -149,7 +150,13 @@ impl Local {
...
+            let copy_result = fs::copy(&temp_path, &dest_path);
+            fs::remove_file(&temp_path)?;
```
Comment (justinas) [inline src/nix/host/local.rs:157]: Defensively remove the temp file even on failure, in order not to leave secrets lying around in the FS. Not sure if necessary.

File: src/nix/key.rs [hunk 4]
```diff
@@ -17,6 +23,18 @@ pub struct Key {
     pub(super) permissions: String,
 }
 
+impl Key {
+    pub(crate) async fn reader(&'_ self,) -> Result<Box<dyn AsyncRead + Send + Unpin + '_>, io::Error> {
```
Comment (justinas) [inline src/nix/key.rs:27]: This function seems very unit-test-able, however I see that there are no unit tests as of now. Should I write one?

File: src/nix/host/local.rs (hunk 2 above)
Comment (zhaofengli) [reply src/nix/host/local.rs:155]: Nice catch!

File: src/nix/key.rs (hunk 1 above)
Comment (zhaofengli) [reply src/nix/key.rs:13]: Yeah, this is definitely nicer and will make it easier to implement #3 in a cleaner way.

File: src/nix/key.rs (hunk 4 above)
Comment (zhaofengli) [reply src/nix/key.rs:27]: Sure, please do! A lot of the interactions with Nix commands can be unit-tested as well, but I haven't got to write them.

**Comment by zhaofengli** [comment]
LGTM, thank you! Let's merge this as is, and we can do more cleanup later. I'll try to make `Key` better typed struct and tackle #3 as well.

**Pull request merged**